hack/generator/pkg/astmodel/property_definition_test.go

```diff
@@ -280,7 +280,7 @@ func Test_PropertyDefinitionMakeOptional_PropertyTypeArrayAndMap(t *testing.T) {
 				original := NewPropertyDefinition(fieldName, fieldJsonName, c.propertyType)
 				field := original.MakeOptional()
 
-				g.Expect(field).To(BeIdenticalTo(original))
+				g.Expect(field).To(Equal(original))
```

Comment:
This change isn't right - the test is checking that the exact same field is returned, avoiding memory churn; merely returning a field that's equal to the original isn't a strong enough assertion.

Reply:
I've fixed this, although:
1. We allocate types quite freely basically everywhere (when modifying immutable stuff, using the visitor, etc) and I don't think this particular location is really even a blip on the radar in terms of how many allocations it drives (because technically speaking it only returns a new copy if the type is an optional Map or optional Array).
2. Including the fix makes the code a bit more awkward IMO because `MakeRequired` and `MakeOptional` end up diverging in how they check for optionality -- `MakeRequired` just needs to know if the type is exactly `Optional` (to unroll it) whereas `MakeOptional` needs to check if it's `Optional`, `Map`, or `Array` (to return early).

I originally hadn't fixed this because it felt like the extra level of awkwardness in the code wasn't worth the negligible perf improvement (especially given how freely we allocate memory elsewhere). I've included the fix in the updated PR though so folks can take a look and let me know if they have a better way to do the checks that's cleaner. If not I left a comment saying why in `MakeOptional`.

Reply:
My motivation is to keep things tidy so that later changes are straightforward. 

For example, in a situation where we're allocating too much memory and we need to increase out sharing of immutable instances, fixes will be easier if our leaf types (like this one) are already optimal with instance reuse.